✨ Added probe for permissive licenses

fhoeborn opened the pull request:
#### What kind of change does this PR introduce?

(Is it a bug fix, feature, docs update, something else?)

- [x] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the current behavior?

#### What is the new behavior (if this is a feature change)?**

- [x] Tests for the changes have been added (for bug fixes/features)

#### Which issue(s) this PR fixes

Fixes #3840

#### Special notes for your reviewer

#### Does this PR introduce a user-facing change?

For user-facing changes, please add a concise, human-readable release note to
the `release-note`

(In particular, describe what changes users might need to make in their
application as a result of this pull request.)

```release-note
Adds a new probe for permissive licenses with the identifier "hasPermissiveLicense"
```

spencerschrock commented:
Can you open an issue for discussion? Our contributing.md outlines the steps:
https://github.com/ossf/scorecard/blob/main/checks/write.md#requirements-for-a-check

> If you'd like to add a check, make sure it meets the following criteria and then create a new GitHub Issue to discuss with the team:

fhoeborn commented:
> Can you open an issue for discussion? Our contributing.md outlines the steps: https://github.com/ossf/scorecard/blob/main/checks/write.md#requirements-for-a-check
> 
> > If you'd like to add a check, make sure it meets the following criteria and then create a new GitHub Issue to discuss with the team:
> 
> (there's also some parts about how to implement that are probably out of date now. but I see you have the implementation using probes which is good)

Thanks for the hint, I opened a issue here https://github.com/ossf/scorecard/issues/3840

cpanato commented on checks/evaluation/permissive_license.go:
```diff
@@ -0,0 +1,79 @@
+// Copyright 2021 OpenSSF Scorecard Authors
```
```suggestion
// Copyright 2024 OpenSSF Scorecard Authors
```

fhoeborn replied:
Thanks, corrected!

cpanato commented on checks/evaluation/permissive_license_test.go:
```diff
@@ -0,0 +1,76 @@
+// Copyright 2023 OpenSSF Scorecard Authors
```
```suggestion
// Copyright 2024 OpenSSF Scorecard Authors
```

fhoeborn replied:
Thanks, corrected!

cpanato commented on checks/permissive_license.go:
```diff
@@ -0,0 +1,60 @@
+// Copyright 2020 OpenSSF Scorecard Authors
```
```suggestion
// Copyright 2024 OpenSSF Scorecard Authors
```

fhoeborn replied:
Thanks, corrected!

cpanato commented on checks/evaluation/permissive_license_test.go:
```diff
@@ -0,0 +1,76 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package evaluation
```
nit: new line

fhoeborn replied:
Thanks, corrected!

cpanato commented on checks/permissive_license_test.go:
```diff
@@ -0,0 +1,105 @@
+// Copyright 2020 OpenSSF Scorecard Authors
```
```suggestion
// Copyright 2024 OpenSSF Scorecard Authors
```

fhoeborn replied:
Thanks, corrected!

cpanato commented on e2e/license_permissive_test.go:
```diff
@@ -0,0 +1,135 @@
+// Copyright 2021 OpenSSF Scorecard Authors
```
```suggestion
// Copyright 2024 OpenSSF Scorecard Authors
```

fhoeborn replied:
Thanks, corrected!

cpanato commented on e2e/license_permissive_test.go:
```diff
@@ -0,0 +1,135 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package e2e
```
nit: new line

fhoeborn replied:
Thanks, corrected!

cpanato commented on probes/hasPermissiveLicense/def.yml:
```diff
@@ -0,0 +1,28 @@
+# Copyright 2023 OpenSSF Scorecard Authors
```
```suggestion
# Copyright 2024 OpenSSF Scorecard Authors
```

fhoeborn replied:
Thanks, corrected!

cpanato commented on probes/hasPermissiveLicense/def.yml:
```diff
@@ -0,0 +1,28 @@
+remediation:
+  effort: High
+  text:
+    - A relicensing is necessary, or the component should not be used by the user for this purpose.
```
nit: new line

fhoeborn replied:
Thanks, corrected!

cpanato commented on probes/hasPermissiveLicense/impl.go:
```diff
@@ -0,0 +1,95 @@
+// Copyright 2023 OpenSSF Scorecard Authors
```
```suggestion
// Copyright 2024 OpenSSF Scorecard Authors
```

fhoeborn replied:
Thanks, corrected!

cpanato commented on probes/hasPermissiveLicense/impl_test.go:
```diff
@@ -0,0 +1,130 @@
+// Copyright 2023 OpenSSF Scorecard Authors
```
```suggestion
// Copyright 2024 OpenSSF Scorecard Authors
```

fhoeborn replied:
Thanks, corrected!

cpanato reviewed (commented):
lgtm

a few nits

spencerschrock commented on probes/hasPermissiveLicense/def.yml:
```diff
@@ -0,0 +1,28 @@
+ The implementation checks whether a permissive license is present
+outcome:
+  - If a license file is found and is permissive, the probe returns a single OutcomePositive.
+  - If a license file is missing the probe returns a single OutcomeNotApplicable.
+  - If the license is not permissive, the probe returns a single OutcomeNegative.
```
Thoughts on no license file also being OutcomeNegative?

fhoeborn replied:
Sounds good to me

spencerschrock commented on probes/hasPermissiveLicense/impl.go:
```diff
@@ -0,0 +1,95 @@
+	if raw == nil {
+		return nil, "", fmt.Errorf("%w: raw", uerror.ErrNil)
+	}
+
+	if raw.LicenseResults.LicenseFiles == nil || len(raw.LicenseResults.LicenseFiles) == 0 {
```
nit/optional: len() works will nil slices, so just the len check should be enough:
```go
if len(raw.LicenseResults.LicenseFiles) == 0 {

}

fhoeborn replied:
Adjusted

spencerschrock commented on probes/hasPermissiveLicense/impl.go:
```diff
@@ -0,0 +1,95 @@
+			// Store the license name in the msg
+			msg := licenseFile.LicenseInformation.Name
+
+			f, err := finding.NewWith(fs, Probe,
+				msg, nil,
+				finding.OutcomePositive)
```
we should have the path of the license file, so including a location would be nice.

fhoeborn replied:
Added path to license file

spencerschrock commented on probes/hasPermissiveLicense/impl.go:
```diff
@@ -0,0 +1,95 @@
+		spdxID := licenseFile.LicenseInformation.SpdxID
+		switch spdxID {
+		case
+			"Unlicense",
+			"Beerware",
```
We have another list of FSF or OSI recognized licenses, not sure I see "Beerware"
https://github.com/ossf/scorecard/blob/e74d90d2164154824de4071073afea4307187be4/checks/raw/license.go#L418

Although it seems our other list is out of date

fhoeborn replied:
Well, the linked list is not only listing permissive licenses (e.g. GPL-3 is on that one)

spencerschrock replied:
this comment was more just a note to self/ project that the list needs updated.

spencerschrock commented on probes/hasPermissiveLicense/impl_test.go:
```diff
@@ -0,0 +1,130 @@
+			},
+		},
+		{
+			name: "nil license files and outcome should be negative",
+			raw: &checker.RawResults{
+				LicenseResults: checker.LicenseData{
+					LicenseFiles: nil,
+				},
+			},
+			outcomes: []finding.Outcome{
+				finding.OutcomeNotApplicable,
```
mismatch between test name and outcome?

spencerschrock commented on probes/hasPermissiveLicense/impl_test.go:
```diff
@@ -0,0 +1,130 @@
+			},
+		},
+		{
+			name: "0 license files and outcome should be negative",
+			raw: &checker.RawResults{
+				LicenseResults: checker.LicenseData{
+					LicenseFiles: []checker.LicenseFile{},
+				},
+			},
+			outcomes: []finding.Outcome{
+				finding.OutcomeNotApplicable,
```
same comment about mismatch

fhoeborn replied:
Adjusted

spencerschrock reviewed (commented):
Reviewed the probe portion, which generally looks good.

fhoeborn commented:
I adjusted the code a bit based upon your review comments, did you have a chance to look into the rest of the PR?

spencerschrock commented on checks/permissive_license.go:
```diff
@@ -0,0 +1,60 @@
+		return checker.CreateRuntimeErrorResult(CheckPermissiveLicense, e)
+	}
+
+	// Set the raw results.
+	pRawResults := getRawResults(c)
+	pRawResults.LicenseResults = rawData
```
note: this would be a race condition between the `License` check and `CheckPermissiveLicense`

spencerschrock commented on probes/entries.go:
```diff
@@ -91,6 +92,9 @@ var (
 		hasLicenseFile.Run,
 		hasFSFOrOSIApprovedLicense.Run,
 	}
+	PermissiveLicense = []ProbeImpl{
+		hasPermissiveLicense.Run,
+	}
```
if this were added to `Uncategorized`, the probe could be called by those interested. even if the check wasn't added

fhoeborn replied:
Done

spencerschrock reviewed (commented):
> I adjusted the code a bit based upon your review comments, did you have a chance to look into the rest of the PR?

The reason I focused on the probe code was based on my comment here: https://github.com/ossf/scorecard/issues/3840#issuecomment-1930698064

I am happy to merge the probe code which is why my review focused on it, but having permissive license as a check is too opinionated (in my opinion) to have as a check for all callers. Full response in the attached issue https://github.com/ossf/scorecard/issues/3840#issuecomment-2033042962

fhoeborn commented:
> > I adjusted the code a bit based upon your review comments, did you have a chance to look into the rest of the PR?
> 
> The reason I focused on the probe code was based on my comment here: [#3840 (comment)](https://github.com/ossf/scorecard/issues/3840#issuecomment-1930698064)
> 
> I am happy to merge the probe code which is why my review focused on it, but having permissive license as a check is too opinionated (in my opinion) to have as a check for all callers. Full response in the attached issue [#3840 (comment)](https://github.com/ossf/scorecard/issues/3840#issuecomment-2033042962)

Thanks for the suggestion!
That works fine for this use case (it is a simple true/false probe anyway), I removed the check code and adjusted the probe to be invocated independently

spencerschrock reviewed (approved):
Thanks for the contribution and being patient / flexible. 

Note: for time reasons I went ahead and added a small bit of metadata (related to #4020) and fixed the linter to get this merged before our v5 prerelease candidate (hopefully tomorrow!).
